pandas/core/groupby/grouper.py

```diff
@@ -668,6 +668,24 @@ def groups(self) -> dict[Hashable, Index]:
         cats = Categorical.from_codes(codes, uniques, validate=False)
         return self._index.groupby(cats)
 
+    @cache_readonly
+    def observed_grouping(self) -> Grouping:
+        if self._observed:
+            return self
```

Comment:
Wouldn't caching `self` prevent it from being garbage collected?

Reply:
Good call - I think a way around it is:

```
@property
def observed_grouping(self) -> Grouping:
    if self._observed:
        return self
    else:
        return self._observed_grouping

@cache_readonly
def _observed_grouping(self) -> Grouping:
    ...
```

Not sure if there is a better pattern here.

Reply:
noted, I have updated the cache method as per comments